### PR TITLE
Mfa multitenancy

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change
(A few sentences about this PR)

## Related issues
- https://github.com/supertokens/supertokens-core/issues/554
- https://github.com/supertokens/supertokens-core/pull/651

## Test Plan
(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes
(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
   - In `build.gradle`
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Item1
- [ ] Item2

### CODE REVIEW

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 1]
```diff
@@ -49,6 +49,9 @@ static String getQueryToCreateTenantConfigsTable(Start start) {
                 + "email_password_enabled BOOLEAN,"
                 + "passwordless_enabled BOOLEAN,"
                 + "third_party_enabled BOOLEAN,"
+                + "totp_enabled BOOLEAN,"
+                + "first_factors TEXT DEFAULT 'null',"
+                + "default_required_factors TEXT DEFAULT 'null',"
```
[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:54]: But im not too sure.. maybe this is fine too.. what are the cons of making them into different tables?

[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:54]: add separate table with hasFirstFactors boolean

[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java [hunk 2]
```diff
@@ -182,6 +182,14 @@ public String getTenantConfigsTable() {
         return addSchemaAndPrefixToTableName("tenant_configs");
     }
 
+    public String getFirstFactorsTable() {
+        return addSchemaAndPrefixToTableName("first_factors");
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java:186]:
```suggestion
        return addSchemaAndPrefixToTableName("tenant_first_factors");
```

File: src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java [hunk 3]
```diff
@@ -182,6 +182,14 @@ public String getTenantConfigsTable() {
...
+    }
+
+    public String getDefaultRequiredFactorIdsTable() {
+        return addSchemaAndPrefixToTableName("default_required_factor_ids");
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java:190]:
```suggestion
        return addSchemaAndPrefixToTableName("tenant_default_required_factor_ids");
```

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 4]
```diff
@@ -122,6 +122,52 @@ public static String getQueryToCreateThirdPartyIdIndexForTenantThirdPartyProvide
...
 
+    public static String getQueryToCreateFirstFactorsTable(Start start) {
+        String schema = Config.getConfig(start).getTableSchema();
+        String tableName = Config.getConfig(start).getFirstFactorsTable();
+        // @formatter:off
+        return "CREATE TABLE IF NOT EXISTS " + tableName + " ("
+                + "connection_uri_domain VARCHAR(256) DEFAULT '',"
+                + "app_id VARCHAR(64) DEFAULT 'public',"
+                + "tenant_id VARCHAR(64) DEFAULT 'public',"
+                + "factor_id VARCHAR(64),"
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:133]:
```suggestion
                + "factor_id VARCHAR(128),"
```

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 5]
```diff
@@ -122,6 +122,52 @@ public static String getQueryToCreateThirdPartyIdIndexForTenantThirdPartyProvide
...
+    public static String getQueryToCreateTenantIdIndexForFirstFactorsTable(Start start) {
+        return "CREATE INDEX IF NOT EXISTS tenant_first_factors_tenant_id_index ON "
+                + getConfig(start).getFirstFactorsTable() + " (connection_uri_domain, app_id, tenant_id);";
+    }
+
+    public static String getQueryToCreateDefaultRequiredFactorIdsTable(Start start) {
+        String schema = Config.getConfig(start).getTableSchema();
+        String tableName = Config.getConfig(start).getDefaultRequiredFactorIdsTable();
+        // @formatter:off
+        return "CREATE TABLE IF NOT EXISTS " + tableName + " ("
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:152]: you need to insert an order column as well per tenant

File: src/main/java/io/supertokens/storage/postgresql/queries/multitenancy/MfaSqlHelper.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public static void createDefaultRequiredFactorIds(Start start, Connection sqlCon, TenantIdentifier tenantIdentifier, String[] defaultRequiredFactorIds)
+            throws SQLException, StorageQueryException {
+        if (defaultRequiredFactorIds == null || defaultRequiredFactorIds.length == 0) {
+            return;
+        }
+
+        String QUERY = "INSERT INTO " + getConfig(start).getDefaultRequiredFactorIdsTable() + "(connection_uri_domain, app_id, tenant_id, factor_id) VALUES (?, ?, ?, ?);";
+        for (String factorId : Set.of(defaultRequiredFactorIds)) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/multitenancy/MfaSqlHelper.java:112]: order matters

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java (hunk 2 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/config/PostgreSQLConfig.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:152]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/multitenancy/MfaSqlHelper.java (hunk 6 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/storage/postgresql/queries/multitenancy/MfaSqlHelper.java]: Done

File: src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java [hunk 7]
```diff
@@ -142,18 +142,19 @@ public static String getQueryToCreateFirstFactorsTable(Start start) {
...
-        String tableName = Config.getConfig(start).getDefaultRequiredFactorIdsTable();
+        String tableName = Config.getConfig(start).getTenantDefaultRequiredFactorIdsTable();
         // @formatter:off
         return "CREATE TABLE IF NOT EXISTS " + tableName + " ("
                 + "connection_uri_domain VARCHAR(256) DEFAULT '',"
                 + "app_id VARCHAR(64) DEFAULT 'public',"
                 + "tenant_id VARCHAR(64) DEFAULT 'public',"
-                + "factor_id VARCHAR(64),"
+                + "factor_id VARCHAR(128),"
+                + "order_idx INTEGER NOT NULL,"
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/storage/postgresql/queries/MultitenancyQueries.java:157]: unique constraint

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Pull request merged**